drivers:dac: added max538x family support

jsanbuen opened the pull request:
added initial driver implementation for max538x family.

For more info: https://www.analog.com/media/en/technical-documentation/data-sheets/MAX5380-MAX5382.pdf

Signed-off-by: Jose Ramon San Buenaventura <jose.sanbuenaventura@analog.com>

buha commented:
also drop the merge commit, it should be just 1 commit with your changes

amiclaus commented on drivers/dac/max538x/max538x.c:
```diff
@@ -0,0 +1,203 @@
+*******************************************************************************/
+int max538x_i2c_transaction(struct max538x_dev *dev, uint8_t *data)
+{
+	uint8_t data_buffer[1];
```
is this extra variable needed?

amiclaus commented on drivers/dac/max538x/max538x.c:
```diff
@@ -0,0 +1,203 @@
+
+	if( (init_param.max538x_vdd == NULL || init_param.max538x_vdd == 0)
+	    && dev->active_device > MAX5381K)
+		return -EINVAL;
```
we should free dev in this case

amiclaus commented on drivers/dac/max538x/max538x.c:
```diff
@@ -0,0 +1,203 @@
+
+	status = no_os_i2c_init(&dev->i2c_desc, &init_param.i2c_init);
+
+	if (status < 0)
```
nit: `if (status)`

amiclaus commented on drivers/dac/max538x/max538x.h:
```diff
@@ -0,0 +1,125 @@
+struct max538x_dev {
+	/* I2C */
+	struct no_os_i2c_desc	*i2c_desc;
+
```
remove all these extra empty lines

amiclaus reviewed (commented):
looks good overall, some minor comments on my side.

amiclaus commented on drivers/dac/max538x/max538x.c:
```diff
@@ -0,0 +1,203 @@
+	if (!dev)
+		return -ENOMEM;
+
+	if( (init_param.max538x_vdd == NULL || init_param.max538x_vdd == 0)
```
`if ((`

amiclaus commented on drivers/dac/max538x/max538x.c:
```diff
@@ -0,0 +1,203 @@
+	dev->active_device = init_param.active_device;
+	init_param.i2c_init.slave_address = chip_info[init_param.active_device].addr;
+
+	if(dev->active_device <= MAX5381K)
```
`if (`

amiclaus commented on drivers/dac/max538x/max538x.c:
```diff
@@ -0,0 +1,203 @@
+{
+	uint8_t vout_code;
+
+	if(vout < 0 || vout > dev->max538x_vref)
```
`if (`

amiclaus commented on drivers/dac/max538x/max538x.h:
```diff
@@ -0,0 +1,125 @@
+/************************** max538x Definitions *******************************/
+/******************************************************************************/
+/* MAX5380 default resolution */
+#define MAX538x_resolution 		0xFF
```
use only uppercase for macro definition

jsanbuen commented:
Changelog/s:
- corrected format based on comments i.e., if ()..
- added "free(dev)" in init condition for inappropriate initial settings.
- changed macro definition for device resolution to all uppercase.
- removed extra variable in max538x_transaction function based on reviewer comment.

jsanbuen commented:
Changelog/s:
-added errno and other libraries to max538x.c and max538x.h files

buha reviewed (commented):
i forgot to publish my review yesterday, i think some of the comments may be outdated

jsanbuen commented:
Changelog/s:
- resolved typo on 5380x (changed to 538x).
- updated dates to 2023(c)
- dropped max538x_i2c_transaction function to use no_os_i2c_write directly.
- changed from malloc to calloc.
- removed whitespaces.

jsanbuen commented:
Changelog/s:
-changed malloc to calloc as recommended on reviewer comment.
-removed whitespaces.
-changed the max5380x to max538x in the comment / header.
-update the date to 2023(c).
-removed max538x_i2c_transaction to use the no_os_i2c_write function directly.

buha commented on drivers/dac/max538x/max538x.c:
```diff
@@ -0,0 +1,188 @@
+
+	return 0;
+
+max538x_init_error:
```
you can use simple labels like:
`error:`

no need to prefix them with anything, their scope is a function so there won't be name clashes

buha commented on drivers/dac/max538x/max538x.c:
```diff
@@ -0,0 +1,188 @@
+	return 0;
+
+max538x_init_error:
+	max538x_remove(dev);
```
best here would be to not use max538x_remove() but manually perform:
```
error_1:
    free(dev);
error:
    no_os_i2c_remove(dev->i2c_desc);
```

And jump to appropriate label with goto.

buha commented on drivers/dac/max538x/max538x.c:
```diff
@@ -0,0 +1,188 @@
+
+	if ((init_param.max538x_vdd == NULL || init_param.max538x_vdd == 0)
+	    && dev->active_device > MAX5381K) {
+		free(dev);
```
(see my other comment below)
here, you could `goto error;`

buha commented on drivers/dac/max538x/max538x.c:
```diff
@@ -0,0 +1,188 @@
+
+	status = no_os_i2c_init(&dev->i2c_desc, &init_param.i2c_init);
+	if (status)
+		goto max538x_init_error;
```
(see my other comment below)
and here, you could `goto error_1;`

jsanbuen commented:
Changelog/s:
-used simplified error labels with goto for error handling in init function.
-changed calloc(0, ....) to calloc(1, ....) based on observed from driver code w/ calloc (ADXL355).

jsanbuen commented:
Changelog/s:
-in init funciton, parameter checking for EINVAL return is placed ahead of other checking.
-corrected goto and label use for when there's error in no_os_i2c_init function based on reviewer comments.

amiclaus commented on drivers/dac/max538x/max538x.c:
```diff
@@ -0,0 +1,185 @@
+ *                            0 - I2C peripheral was initialized and the
+ *                                device is present.
+*******************************************************************************/
+int32_t max538x_init(struct max538x_dev **device,
```
nit: we switched to using standard `int` for the function types. please update them accordingly (like you did for `int max538x_set_voutput(struct max538x_dev *dev, float vout)`)

jsanbuen commented:
Changelog/s
-changed function types to int from int32_t for init and remove functions.

amiclaus commented on drivers/dac/max538x/max538x.c:
```diff
@@ -0,0 +1,185 @@
+*******************************************************************************/
+int max538x_remove(struct max538x_dev *dev)
+{
+	int32_t ret;
```
`ret` type should match the function type.

amiclaus commented on drivers/dac/max538x/max538x.c:
```diff
@@ -0,0 +1,185 @@
+		 struct max538x_init_param init_param)
+{
+	struct max538x_dev *dev;
+	int32_t status;
```
`status` type should match the function type. Also consider using throughout the code only one naming: either `status`, either `ret`

buha commented:
in order to merge, @jsanbuen please fix the cppcheck and remaining comments from @amiclaus

jsanbuen commented:
Changelog/s:
- changed int32_t types to int.
- consistent use of variable ret for all other functions.

jsanbuen commented:
Changelog/s:
-in init function, corrected the initial EINVAL check by changing dev->active_device to init_param.active device. Checked cppcheck in pipeline and this part was initially source of error in cppcheck.
